### ISSUE TITLE
can u tell me ur rate of recognition?

### ISSUE THREAD

**Issue by gkyustc** [thnkim/OpenFacePytorch#4]
thank you very much

**Comment by thnkim** [comment]
hello, I don't have the data now, checked a long time ago.
Do you have an issue that the result are different from the original OpenFace torch?
If so, I'll see it soon (I'm on biz trip now). More detailed information (e.g., failure cases) will be very helpful.
Thank you.

**Comment by gkyustc** [comment]
I just test your model on lfw dataset, the recognition is about 60%. So I don't think your 'pth' model is as same as 'nn4.small2.v1.t7'.
well, when I tried to use **load_lua** to load the 'nn4.small2.v1.t7' model, I cannot use **load_state_dict**. Have you ever used the original 'nn4.small2.v1.t7' to do some pretrained work?

**Comment by thnkim** [comment]
@gky1993 
Thank you so much for the report. 
I've just found that I have two nn4.small2.v1.t7 of which md5sum are different. I guess it is the possible cause.
The conversion process was complicated: t7 to hdf5 and so on. I'll debug it.

**Comment by thnkim** [comment]
@gky1993 
I tested my code to extract all features for LFW data (exactly the same files that openface uses, referring to https://cmusatyalab.github.io/openface/models-and-accuracies/#accuracy-on-the-lfw-benchmark).
Comparing 13,233 features extracted from the files, I got a list of MSE's (== ||128d_feature_from_openface - 128d_feature_from_mine||^2 / 128). 
Its mean, min, and max are 1.3736351296200752e-07, 9.39600558765545e-13, and 0.00012538873907050677, respectively.
When I print out numpy.histogram() of the error list, I got
```
(array([13181,    23,    14,     6,     5,     2,     0,     1,     0,     1]), array([  9.39600559e-13,   1.25388748e-05,   2.50777486e-05,
         3.76166224e-05,   5.01554962e-05,   6.26943700e-05,
         7.52332438e-05,   8.77721176e-05,   1.00310991e-04,
         1.12849865e-04,   1.25388739e-04]))
```

Ok. I agree that the output from mine does not guarantee 100% reproduction of OpenFace maybe due to a numerical error during the conversion. 
But, I don't understand why you got 60% of accuracy.
I guess, either files in this repo are different from my working codes (I'll double-check) or face recognition for LFW is very sensitive to the scale of errors, under the assumption that your test is surely correct.

**Comment by gkyustc** [comment]
I test your model with the code I used to test center loss which accuracy is about 98%.  So I don't think there is something wrong with my verification code. In case that I extracted wrong features, I show you one person's feature:
**name**:
AJ_Cook/AJ_Cook_0001.jpg

**feature**：0.0543527 0.197927 -0.0498536 -0.0377145 0.0492867 0.23241 -0.02947 -0.12735 -0.115022 0.0489607 0.00862604 -0.0416681 0.0617135 0.0263894 0.117972 -0.0888898 0.0186169 0.00439241 0.0392823 -0.0466764 -0.033479 0.11938 -0.0817543 -0.0375794 0.0747962 0.0117022 -0.090453 -0.0290205 0.033467 0.107535 0.0258363 0.0950073 -0.107649 -0.0147895 0.0502639 0.0691447 0.0535041 -0.103016 0.139346 0.0338923 0.100073 -0.077405 0.00600538 -0.0292218 -0.106833 0.157141 0.103322 0.0269465 -0.106955 -0.119268 0.109367 0.164977 0.0389948 0.0241409 -0.0105076 0.0132947 0.0194844 -0.0030958 0.0557375 -0.0659023 -0.0649719 0.062567 0.0128209 -0.111545 -0.0389141 0.00851676 0.149448 0.107717 0.027667 0.058154 0.102781 0.073389 -0.0283427 -0.0958955 0.0996606 0.0905334 -0.179241 -0.00912448 -0.0203346 0.111051 0.0526506 0.151108 0.113085 0.0932733 0.0317884 0.0716961 0.079071 -0.0771614 -0.0935161 0.0309819 -0.0275139 -0.0765612 0.0423179 0.042763 -0.204559 -0.175942 0.0683432 0.0268036 0.0254427 -0.0931672 0.0506866 0.176417 0.0648923 0.105888 -0.17106 0.0479559 -0.057369 0.116307 -0.091246 0.0844345 0.10889 -0.0818939 -0.131961 -0.0203142 -0.0744118 -0.0222884 -0.228583 -0.0958232 -0.0658109 -0.019378 0.120314 -0.0982558 0.0262307 -0.0252392 0.0693105 0.0429171 -0.00625768 0.0577717 

Since  you have extract the features from LFW now, can you tell me your rate of recognition?

**Comment by thnkim** [comment]
I'm not afford to implement LFW benchmark now, due to my current project.
What I can do now is dump the features.

BTW, the feature that I extracted are very different than yours.  Features I got were:
OpenFace (torch7):
```
[0.039907478, -0.10562486, 0.017258899, -0.057111528, -0.044686045, 0.10753872, -0.026802914, 0.13901976, 0.08700113, -0.14811194, -0.10012631, 0.075193383, -0.048805851, -0.11999057, -0.084125593, -0.026385071, -0.18341038, -0.12564693, -0.13456397, 0.020642072, 0.027960822, 0.090467408, 0.11452021, -0.1228845, -0.20560969, 0.035669345, 0.007906111, -0.12346546, -0.19183068, -0.10784543, -0.057633758, -0.024470106, 0.038534094, 0.10723812, -0.019461472, -0.018667607, 0.0092111295, 0.15282167, -0.16931115, 0.059797719, -0.042360395, -0.0048624855, -0.13485095, -0.038039733, -0.0036135614, -0.011970996, -0.031772681, -0.073736288, -0.039606169, 0.27637151, 0.01472572, -0.029435562, -0.024644274, 0.14058885, 0.1028952, -0.079395041, 0.093963541, -0.069812499, -0.080632448, -0.052175142, 0.014416496, 0.0061943233, 0.033850692, -0.026692104, 0.078019544, -0.18845877, -0.0022525699, 0.10216717, 0.024317836, 0.16037147, -0.00063781312, -0.065962769, 0.012540579, 0.086328678, -0.061900172, 0.024730388, -0.032168079, -0.1243391, -0.014312956, 0.022302518, -0.0080400705, -0.054967362, -0.013650517, -0.012844076, -0.03439283, -0.08138717, 0.065033823, 0.062611587, 0.042419285, 0.095196843, 0.036855131, -0.10851701, -0.061664443, 0.021430861, -0.0047197733, 0.12309733, 0.08157777, -0.036871377, -0.044295911, 0.11677667, 0.065086216, 0.025524082, -0.13825029, -0.060443528, 0.046535332, -0.1104449, -0.10598736, -0.071873948, -0.13145982, -0.015212451, 0.017581131, 0.089625902, 0.015027259, -0.04751711, 0.075157195, 0.078594133, 0.079592109, -0.044946954, 0.059139531, 0.054601613, 0.057257142, 0.07763242, 0.17033957, -0.028281284, 0.19039896, -0.076656312, 0.17893907, 0.011994334]
```
My code (PyTorch):
```
[0.03990338370204, -0.10562666505575, 0.01726321130991, -0.057117193937302, -0.044686481356621, 0.10753782838583, -0.026803629472852, 0.13902050256729, 0.087006442248821, -0.14811037480831, -0.10012611746788, 0.075193598866463, -0.048803813755512, -0.11998648196459, -0.084124647080898, -0.026386052370071, -0.18341076374054, -0.12564630806446, -0.13457031548023, 0.02064037322998, 0.027959449216723, 0.090468376874924, 0.11451844871044, -0.12288132309914, -0.20561255514622, 0.035672139376402, 0.0079063950106502, -0.12346003204584, -0.19182927906513, -0.10785017162561, -0.05762929096818, -0.02446972951293, 0.038533706218004, 0.1072396710515, -0.01946390606463, -0.018666081130505, 0.0092099132016301, 0.15282678604126, -0.16931068897247, 0.059801891446114, -0.042358703911304, -0.0048604928888381, -0.13485442101955, -0.038043387234211, -0.0036098889540881, -0.011978446505964, -0.03177884221077, -0.073732480406761, -0.0396025814116, 0.27637043595314, 0.014721915125847, -0.029438579455018, -0.024645734578371, 0.14058808982372, 0.10289344191551, -0.079393059015274, 0.093960210680962, -0.069806598126888, -0.080635957419872, -0.05217632651329, 0.014413501136005, 0.0061954809352756, 0.033851984888315, -0.02668958902359, 0.07801578193903, -0.1884531378746, -0.0022514925803989, 0.10217117518187, 0.024316085502505, 0.16037157177925, -0.00063659477746114, -0.065963760018349, 0.012541764415801, 0.086332403123379, -0.061897914856672, 0.024728838354349, -0.032168854027987, -0.12434236705303, -0.014311259612441, 0.022298270836473, -0.0080393077805638, -0.054964683949947, -0.013647451065481, -0.012845057062805, -0.034394882619381, -0.081390634179115, 0.065038345754147, 0.062615893781185, 0.042422611266375, 0.095195487141609, 0.036856085062027, -0.1085190474987, -0.061657357960939, 0.021431950852275, -0.0047166268341243, 0.12309870123863, 0.081573717296124, -0.036874681711197, -0.044297862797976, 0.11677874624729, 0.065086089074612, 0.025519641116261, -0.13825498521328, -0.060441818088293, 0.046537786722183, -0.11044345796108, -0.10599082708359, -0.07188156247139, -0.13145849108696, -0.015215691179037, 0.01758143492043, 0.089630223810673, 0.015027099289, -0.047511722892523, 0.075158476829529, 0.078594267368317, 0.079590864479542, -0.044944077730179, 0.059139624238014, 0.054600961506367, 0.057259280234575, 0.077631689608097, 0.1703377366066, -0.028281815350056, 0.19039416313171, -0.076655149459839, 0.1789333075285, 0.011993756517768]
```

Question #1:
Also, the file that I used to reproduce 'OpenFace's was openface/data/lfw/dlib-affine-sz:96/AJ_Cook/AJ_Cook_0001.png', not .jpg file. Do I use the dataset different than yours?

Question #2:
What is the features you got from 1) OpenFace and 2) my code? The feature you attached was 2), then how about 1)?

Question #3:
Also can you attach your 'AJ_Cook/AJ_Cook_0001.jpg'?

Question #4:
I guess, you might use images not aligned. My code is just to extract features for the recognition. The input image is assumes to be cropped and aligned. OpenFace's scope is bigger than face recognition itself, so it contains face detection and alignment also. Is my guess right or you got the weird result even though you use the aligned face images?

**Comment by shayansiddiqui** [comment]
@thnkim 
Have you verified that the model on this repo is the correct one or not?

Thanks

**Comment by gkyustc** [comment]
I'm sorry that I'm a little busy recently. I didn't test original openface's model, I just tested yours.  the dataset I used is aligned. The dataset  on our lab's server, not downloded by me  , here is the image 
![image](https://user-images.githubusercontent.com/20548130/35078423-2e657876-fc3d-11e7-9b6c-673fe0237d31.png)

**Comment by thnkim** [comment]
@shayansiddiqui @gky1993 
I've just uploaded newly converted model file: openface_20180119.pth. This should work.

BTW, @gky1993, The AJ_Cook_0001.png that openface (torch7 version) generated is different than yours:
![aj_cook_0001](https://user-images.githubusercontent.com/7189259/35107324-fe60cb90-fcb3-11e7-9c56-6e4ef5490851.png)

I followed up 'Running The LFW Experiment' in https://cmusatyalab.github.io/openface/models-and-accuracies/#accuracy-on-the-lfw-benchmark.

ps. Sorry for the belated response, I'm busy recently due to my urgent project :(

**Comment by botcs** [comment]
@thnkim  Thank you for transforming the network weights and writing the equivalent of OpenFace!

I think the problem is the complete lack of preprocess in the PyTorch transcription.

This can be tested with any image I guess: if you crop any image the distance will be far more large between the original and the cropped than other images.

Could you reimplement the preprocessing steps of OpenFace?

**Comment by thnkim** [comment]
@botcs Let me see. The first intention was to convert OpenFace's neural network part only, but I'll see if the preprocessing is simple. (It should be, I guess. They are based on dlib.)

**Comment by botcs** [comment]
I managed to redo all the alignment that the original OpenFace lib was doing, and now it works super-efficient with your translation. Amazing stuff! Thank you!
